File: javascript/canConstruct.js
const canConstruct = (target, wordBank, memo = {}) => {
    if (target in memo) return memo[target];
    if (target === '') {
        return true;
    }

    for (let word of wordBank) {
        if (target.indexOf(word) === 0 ) {
            const suffix = target.slice(word.length);
            if (canConstruct(suffix, wordBank, memo) === true) {
            memo[target] = true;
            return true;
            }
        }
    }
    memo[target] = false;
    return false;
}

console.log(canConstruct("abcdefsdadas", ["a", "b", "d", "s.3", "f", "cdef"]));

// Brute Force 
// Time O(n^m * m)
// Space O(m^2)

// memo
// Time O(n*m^2)
// Space O(m^2)

File: ruby/memoization/canConstruct.rb
def can_construct(target, wordBank)
return true if target === ''

    wordBank.each do |word|
        
        if target.index(word) === 0
            suffix = target.slice(word.size, target.size)
            return true if can_construct(suffix, wordBank) === true
        
        end
    end

    return false
end

p can_construct("abcdefsdadas", ["abcdef", "b", "da", "s", "ff", "cdef"])